src/js/components/FileInput/FileInput.js

```diff
@@ -0,0 +1,257 @@
+import { Keyboard } from '../Keyboard';
+import { Text } from '../Text';
+
+import { StyledFileInput } from './StyledFileInput';
+
+// We want the interaction of <input type="file" /> but none of its styling.
+// So, we put what we want to show underneath and
+// position the <input /> on top with an opacity of zero.
+// If there are any files selected, we need to show the buttons to remove them.
+// So, we offset the <input /> from the right by the appropriate width.
+// We don't use Stack because of how we need to control the positioning.
+
+const ContentsBox = styled(Box)`
+  position: relative;
+  ${props => (props.disabled ? disabledStyle() : 'cursor: pointer;')}
+  &:focus {
+    ${focusStyle}
+  }
+  ${props => props.theme.fileInput && props.theme.fileInput.extend};
+  ${props =>
+    props.hover &&
+    props.theme.fileInput &&
+    props.theme.fileInput.hover &&
+    props.theme.fileInput.hover.extend};
+  ${props =>
+    props.dragOver &&
+    props.theme.fileInput &&
+    props.theme.fileInput.dragOver &&
+    props.theme.fileInput.dragOver.extend};
+`;
+
+const Label = styled(Text)`
+  ${props =>
+    props.theme.fileInput &&
+    props.theme.fileInput.label &&
+    props.theme.fileInput.label.extend};
+`;
+
+const Message = styled(Text)`
+  ${props =>
+    props.theme.fileInput &&
+    props.theme.fileInput.message &&
+    props.theme.fileInput.message.extend};
+`;
+
+const FileInput = forwardRef(
+  (
+    {
+      a11yTitle,
+      background,
+      border,
+      disabled,
+      id,
+      renderFile,
+      messages,
+      margin,
+      multiple,
+      name,
+      onChange,
+      pad,
+      value: valueProp,
+      ...rest
+    },
+    ref,
+  ) => {
+    const theme = useContext(ThemeContext);
+    const formContext = useContext(FormContext);
+    const [files, setFiles] = formContext.useFormInput(name, valueProp, []);
+    const [hover, setHover] = React.useState();
+    const [dragOver, setDragOver] = React.useState();
+    const aggregateThreshold = (multiple && multiple.aggregateThreshold) || 10;
+    const inputRef = useForwardedRef(ref);
+    const controlRef = useRef();
+    const removeRef = useRef();
+    const RemoveIcon = theme.fileInput.icons.remove;
+
+    const mergeTheme = (propertyName, defaultKey) => {
+      let result = {};
+      const themeProp = theme.fileInput[propertyName];
+      if (themeProp)
+        if (typeof themeProp !== 'object')
+          if (defaultKey) result[defaultKey] = themeProp;
+          else result = themeProp;
+        else result = { ...themeProp };
+      const hoverThemeProp = theme.fileInput.hover[propertyName];
+      if (hover && hoverThemeProp)
+        if (typeof hoverThemeProp !== 'object')
+          if (defaultKey) result[defaultKey] = hoverThemeProp;
+          else result = hoverThemeProp;
+        else result = { ...result, ...hoverThemeProp };
+      const dragOverThemeProp = theme.fileInput.dragOver[propertyName];
+      if (dragOver && dragOverThemeProp)
```

Comment:
Not sure how I feel about the border becoming red when I am dragging a file over. I thought it was giving me an error state, maybe that I couldn't drop that kind of file or something.

Reply:
Oops, good catch. Changed to "control".